Changes to prepare for fake tensors on in functorch by default

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84447
* #84671
* __->__ #84432

Fixes some errors you run into in dynamo when turning on fake tensors. I'm waiting on flipping the switch because I need to also get some fixes into dynamo + do benchmarking.

I could manually turn off fake tensors in functorch in dynamo, and then turn it on here if requested, although the changes here are pretty minimal.

eellison committed:
Switch on fake tensors in functorch by default

[ghstack-poisoned]

eellison committed:
Update on "Switch on fake tensors in functorch by default"




[ghstack-poisoned]

eellison committed:
Update on "Switch on fake tensors in functorch by default"




[ghstack-poisoned]

eellison committed:
Update on "Switch on fake tensors in functorch by default"




[ghstack-poisoned]

eellison committed:
Update on "Switch on fake tensors in functorch by default"




[ghstack-poisoned]

eellison committed:
Update on "Switch on fake tensors in functorch by default"




[ghstack-poisoned]

Chillee commented on torch/fx/passes/backends/cudagraphs.py:
```diff
@@ -21,15 +22,17 @@ def is_node_supported(self, submodules, node: torch.fx.Node) -> bool:
 
         found_not_cuda = False
 
+        FK = "val" if functorch.compile.config.use_fake_tensor else "fake_result"
```
I think this should always just be "val" now.

anijain2305 replied:
This will also fix the `aot_cudagraphs` 0% passrate observed in the nightlies.

Chillee replied:
We should just fix all the uses of `fake_result` to use `val` instead.

eellison replied:
If we have `aot_cudagraphs` at 0% in nightlies we should temporarily disable the tests

eellison replied:
im just going to do a polyfill for now and we can update later

Chillee commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -843,9 +843,7 @@ def wrap(e, device=None):
-        return (
-            self.const_tensors and t.numel() <= CONSTANT_NUMEL_LIMIT and not t.is_sparse
-        )
+        return self.const_tensors and t.numel() <= CONSTANT_NUMEL_LIMIT and not t.is_sparse and not isinstance(t, FakeTensor)
```
Why is this?

eellison replied:
you don't want to make a constant when you run an fx graph, and the input is a fake tensor (obviously that's not a constant)

Chillee reviewed (approved):
LGTM

eellison committed:
Update on "Changes to prepare for fake tensors on in functorch by default"


Fixes some errors you run into in dynamo when turning on fake tensors. I'm waiting on flipping the switch because I need to also get some fixes into dynamo + do benchmarking.

I could manually turn off fake tensors in functorch in dynamo, and then turn it on here if requested, although the changes here are pretty minimal. 

[ghstack-poisoned]

eellison committed:
Update on "Changes to prepare for fake tensors on in functorch by default"


Fixes some errors you run into in dynamo when turning on fake tensors. I'm waiting on flipping the switch because I need to also get some fixes into dynamo + do benchmarking.

I could manually turn off fake tensors in functorch in dynamo, and then turn it on here if requested, although the changes here are pretty minimal. 

[ghstack-poisoned]

eellison committed:
Update on "Changes to prepare for fake tensors on in functorch by default"


Fixes some errors you run into in dynamo when turning on fake tensors. I'm waiting on flipping the switch because I need to also get some fixes into dynamo + do benchmarking.

I could manually turn off fake tensors in functorch in dynamo, and then turn it on here if requested, although the changes here are pretty minimal. 

[ghstack-poisoned]

eellison commented:
@pytorchbot merge
